### PR TITLE
fix: fullURL endpoint generation

### PR DESCRIPTION
**Pull request by eiixy**
fix: generation azure fullURL #794
feat: Use setters pattern for fullURL method #636

### CODE REVIEW

File: api_internal_test.go [hunk 1]
```diff
@@ -153,20 +164,34 @@ func TestCloudflareAzureFullURL(t *testing.T) {
...
 			"https://gateway.ai.cloudflare.com/v1/dnekeim2i39dmm4mldemakiem3i4mkw3/demo/azure-openai/resource/chatgpt-demo/" +
 				"chat/completions?api-version=2023-05-15",
 		},
+		{
+			"",
+			"https://gateway.ai.cloudflare.com/v1/dnekeim2i39dmm4mldemakiem3i4mkw3/demo/azure-openai/resource/",
+			"/assistants?limit=10",
+			"chatgpt-demo",
+			"https://gateway.ai.cloudflare.com/v1/dnekeim2i39dmm4mldemakiem3i4mkw3/demo/azure-openai/resource" +
+				"/assistants?api-version=2023-05-15&limit=10",
```
Comment (eiixy) [inline api_internal_test.go:193]: This test case needs further verification. Currently, I haven't found additional documentation.
https://developers.cloudflare.com/ai-gateway/providers/azureopenai/

File: edits.go [hunk 2]
```diff
@@ -38,7 +38,10 @@ will need to migrate to GPT-3.5 Turbo by January 4, 2024.
 You can use CreateChatCompletion or CreateChatCompletionStream instead.
 */
 func (c *Client) Edits(ctx context.Context, request EditsRequest) (response EditsResponse, err error) {
-	req, err := c.newRequest(ctx, http.MethodPost, c.fullURL("/edits", fmt.Sprint(request.Model)), withBody(request))
+	req, err := c.newRequest(ctx, http.MethodPost, c.fullURL(
```
Comment (sashabaranov) [inline edits.go:41]: nit: could we please make it one argument per line?

Comment (sashabaranov) [reply edits.go:41]: (applies to some other files as well)

File: client.go [hunk 3]
```diff
@@ -222,42 +222,54 @@ func decodeString(body io.Reader, output *string) error {
...
+	if c.config.APIType == APITypeAzure || c.config.APIType == APITypeAzureAD {
+		azureDeploymentName := c.config.GetAzureDeploymentByModel(args.model)
+		if azureDeploymentName == "" {
+			azureDeploymentName = "UNKNOWN"
+		}
+		baseURL = fmt.Sprintf("%s/%s", baseURL, azureAPIPrefix)
+		if containsSubstr(azureDeploymentsEndpoints, suffix) {
+			baseURL = fmt.Sprintf("%s/%s/%s", baseURL, azureDeploymentsPrefix, azureDeploymentName)
+		}
+	}
```
Comment (sashabaranov) [inline client.go:271]: This part is pretty hard to figure about 🤔  Could I suggest breaking it down into more functions and keeping `fullURLOptions` closer to the code that uses it?

Something like this:

```go
func (c *Client) suffixWithAPIVersion(previousSuffix string) (newSuffix string) {
	parsedSuffix, err := url.Parse(previousSuffix)
	if err != nil {
		panic("failed to parse url suffix")
	}

	query := parsedSuffix.Query()
	query.Add("api-version", c.config.APIVersion)
	return fmt.Sprintf("%s?%s", parsedSuffix.Path, query.Encode())
}

func (c *Client) baseURLWithAzureDeployment(baseURL, suffix, model string) (newBaseURL string) {
	azureDeploymentName := c.config.GetAzureDeploymentByModel(model)
	if azureDeploymentName == "" {
		azureDeploymentName = "UNKNOWN"
	}
	baseURL = fmt.Sprintf("%s/%s", baseURL, azureAPIPrefix)
	if containsSubstr(azureDeploymentsEndpoints, suffix) {
		baseURL = fmt.Sprintf("%s/%s/%s", baseURL, azureDeploymentsPrefix, azureDeploymentName)
	}
	return baseURL
}


// fullURL returns full URL for request.
func (c *Client) fullURL(suffix string, setters ...fullURLOption) string {
	baseURL := strings.TrimRight(c.config.BaseURL, "/")
	urlOptions := fullURLOptions{}
	for _, setter := range setters {
		setter(&args)
	}

	if c.config.APIType == APITypeAzure || c.config.APIType == APITypeAzureAD {
		baseURL = baseURLWithAzureDeployment(baseURL, suffix, urlOptions.model)
	}

	if c.config.APIVersion != "" {
		suffix = c.suffixWithAPIVersion(suffix)
	}

	return fmt.Sprintf("%s%s", baseURL, suffix)
}
```

Comment (eiixy) [reply client.go:283]: I've completed the suggested changes. Could you please review it again?

**Review by sashabaranov — APPROVED** [review]
Great work, thank you!

**Pull request merged**